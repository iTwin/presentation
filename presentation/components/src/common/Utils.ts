/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

import * as React from "react";
import { I18NNamespace } from "@bentley/imodeljs-i18n";
import { Presentation } from "@bentley/presentation-frontend";

/**
 * An interface of something that has a priority.
 * @internal
 */
export interface IPrioritized {
  priority: number;
}

/**
 * A sorting algorithm for `Array.sort` that sorts items by priority.
 * Higher priority items appear first in the list.
 *
 * @internal
 */
export const prioritySortFunction = (a: IPrioritized, b: IPrioritized): number => {
  if (a.priority > b.priority)
    return -1;
  if (a.priority < b.priority)
    return 1;
  return 0;
};

let localizationNamespace: I18NNamespace | undefined;
/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = async (stringId: string): Promise<string> => {
  const localizationNamespaceName = "PresentationComponents";
  if (!localizationNamespace) {
    localizationNamespace = Presentation.i18n.registerNamespace(localizationNamespaceName);
  }
  await localizationNamespace.readFinished;
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.i18n.translate(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName)
    return component.displayName;
  if (component.name)
    return component.name;
  return "Component";
};
