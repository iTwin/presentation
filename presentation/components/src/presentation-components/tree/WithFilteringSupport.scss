/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

.filteredTree {

  position: relative;
  width: 100%;
  height: 100%;

  .filteredTreeOverlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 9999;
    background-color: rgba(255,255,255,.8);
  }

}
