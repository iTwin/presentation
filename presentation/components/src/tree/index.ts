/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

export { default as PresentationTreeDataProvider } from "./DataProvider";
export { default as IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
export { default as withUnifiedSelection, SelectionTarget } from "./WithUnifiedSelection";
export { default as withFilteringSupport } from "./WithFilteringSupport";
