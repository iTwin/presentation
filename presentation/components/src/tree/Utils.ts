/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import { StyleHelper } from "../common/StyleHelper";
import { CheckBoxState } from "@bentley/ui-core";
import { Node, PageOptions as PresentationPageOptions, LabelDefinition } from "@bentley/presentation-common";
import { DelayLoadedTreeNodeItem, PageOptions as UiPageOptions, ItemStyle, ItemColorOverrides } from "@bentley/ui-components";
import { createLabelRecord } from "../common/Utils";

/** @internal */
export const PRESENTATION_TREE_NODE_KEY = "__presentation-components/key";

/** @internal */
export const createTreeNodeItems = (nodes: ReadonlyArray<Readonly<Node>>, parentId?: string): DelayLoadedTreeNodeItem[] => {
  const list = new Array<DelayLoadedTreeNodeItem>();
  for (const node of nodes)
    list.push(createTreeNodeItem(node, parentId));
  return list;
};

/** @internal */
export const createTreeNodeItem = (node: Readonly<Node>, parentId?: string): DelayLoadedTreeNodeItem => {
  const item: DelayLoadedTreeNodeItem = {
    id: [...node.key.pathFromRoot].reverse().join("/"),
    label: node.label,
    labelDefinition: createLabelRecord(node.labelDefinition ? node.labelDefinition : LabelDefinition.fromLabelString(node.label), "node_label"),
  };
  (item as any)[PRESENTATION_TREE_NODE_KEY] = node.key;

  const style: ItemStyle = {};
  const colorOverrides: ItemColorOverrides = {};
  if (parentId)
    item.parentId = parentId;
  if (node.description)
    item.description = node.description;
  if (node.hasChildren)
    item.hasChildren = true;
  if (node.isExpanded)
    item.autoExpand = true;
  if (node.imageId)
    item.icon = node.imageId;
  if (StyleHelper.isBold(node))
    style.isBold = true;
  if (StyleHelper.isItalic(node))
    style.isItalic = true;
  const foreColor = StyleHelper.getForeColor(node);
  if (foreColor)
    colorOverrides.color = foreColor;
  const backColor = StyleHelper.getBackColor(node);
  if (backColor)
    colorOverrides.backgroundColor = backColor;
  if (node.isCheckboxVisible) {
    item.isCheckboxVisible = true;
    if (node.isChecked)
      item.checkBoxState = CheckBoxState.On;
    if (!node.isCheckboxEnabled)
      item.isCheckboxDisabled = true;
  }
  if (Object.keys(colorOverrides).length > 0)
    style.colorOverrides = colorOverrides;
  if (Object.keys(style).length > 0)
    item.style = style;
  if (node.extendedData)
    item.extendedData = node.extendedData;

  return item;
};

/** @internal */
export const pageOptionsUiToPresentation = (pageOptions?: UiPageOptions): PresentationPageOptions | undefined => {
  if (pageOptions)
    return { ...pageOptions };
  return undefined;
};
