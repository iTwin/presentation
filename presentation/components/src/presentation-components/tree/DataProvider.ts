/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import memoize from "micro-memoize";
import { IDisposable, Logger } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { HierarchyRequestOptions, NodeKey, NodePathElement, Ruleset } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, PageOptions, TreeNodeItem } from "@bentley/ui-components";
import { RulesetRegistrationHelper } from "../common/RulesetRegistrationHelper";
import { IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
import { createTreeNodeItems, pageOptionsUiToPresentation, PRESENTATION_TREE_NODE_KEY } from "./Utils";

/**
 * Properties for creating a `PresentationTreeDataProvider` instance.
 * @public
 */
export interface PresentationTreeDataProviderProps {
  /** IModel to pull data from. */
  imodel: IModelConnection;
  /** Id of the ruleset to use when requesting content or a ruleset itself. */
  ruleset: string | Ruleset;
  /**
   * Paging size for obtaining nodes.
   *
   * Presentation data providers, when used with paging, have ability to save one backend request for size / count. That
   * can only be achieved when `pagingSize` property is set on the data provider and it's value matches size which is used when
   * requesting nodes. To help developers notice this problem, data provider emits a warning similar to this:
   * ```
   * PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call. Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.
   * ```
   * To fix the issue, developers should make sure the page size used for requesting data is also set for the data provider:
   * ```TS
   * const pagingSize = 10;
   * const provider = new TreeDataProvider({imodel, ruleset, pagingSize});
   * // only one backend request is made for the two following requests:
   * provider.getNodesCount();
   * provider.getNodes({ start: 0, size: pagingSize });
   * ```
   */
  pagingSize?: number;
}

/**
 * Presentation Rules-driven tree data provider.
 * @public
 */
export class PresentationTreeDataProvider implements IPresentationTreeDataProvider, IDisposable {
  private _imodel: IModelConnection;
  private _rulesetRegistration: RulesetRegistrationHelper;
  private _pagingSize?: number;

  /** Constructor. */
  public constructor(props: PresentationTreeDataProviderProps) {
    this._rulesetRegistration = new RulesetRegistrationHelper(props.ruleset);
    this._imodel = props.imodel;
    this._pagingSize = props.pagingSize;
  }

  /** Destructor. Must be called to clean up.  */
  public dispose() {
    this._rulesetRegistration.dispose();
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetRegistration.rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get imodel(): IModelConnection { return this._imodel; }

  /**
   * Paging options for obtaining nodes.
   * @see `PresentationTreeDataProviderProps.pagingSize`
   */
  public get pagingSize(): number | undefined { return this._pagingSize; }
  public set pagingSize(value: number | undefined) { this._pagingSize = value; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._imodel,
      rulesetOrId: this._rulesetRegistration.rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return (node as any)[PRESENTATION_TREE_NODE_KEY];
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public async getNodes(parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> {
    if (undefined !== pageOptions && pageOptions.size !== this.pagingSize) {
      const msg = `PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call.
        Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.`;
      Logger.logWarning("Presentation.Components", msg);
    }

    return (await this._getNodesAndCount(parentNode, pageOptions)).nodes;
  }

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public async getNodesCount(parentNode?: TreeNodeItem): Promise<number> {
    const pageOptions = undefined !== this.pagingSize ? { start: 0, size: this.pagingSize } : undefined;
    return (await this._getNodesAndCount(parentNode, pageOptions)).count;
  }

  private _getNodesAndCount = memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<{ nodes: TreeNodeItem[], count: number }> => {
    const requestCount = undefined !== pageOptions && 0 === pageOptions.start && undefined !== pageOptions.size;
    const parentKey = parentNode ? this.getNodeKey(parentNode) : undefined;

    if (!requestCount) {
      const allNodes = await Presentation.presentation.getNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return { nodes: parentNode ? createTreeNodeItems(allNodes, parentNode.id) : createTreeNodeItems(allNodes), count: allNodes.length };
    }

    const nodesResponse = await Presentation.presentation.getNodesAndCount({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
    return { nodes: parentNode ? createTreeNodeItems(nodesResponse.nodes, parentNode.id) : createTreeNodeItems(nodesResponse.nodes), count: nodesResponse.count };
  }, { isMatchingKey: MemoizationHelpers.areNodesRequestsEqual as any });

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha Hierarchy loading performance needs to be improved before this becomes publicly available.
   */
  public async loadHierarchy() {
    return Presentation.presentation.loadHierarchy(this.createRequestOptions());
  }

}

class MemoizationHelpers {
  public static areNodesRequestsEqual(lhsArgs: [TreeNodeItem?, PageOptions?], rhsArgs: [TreeNodeItem?, PageOptions?]): boolean {
    if (lhsArgs[0]?.id !== rhsArgs[0]?.id)
      return false;
    if ((lhsArgs[1]?.start ?? 0) !== (rhsArgs[1]?.start ?? 0))
      return false;
    if ((lhsArgs[1]?.size ?? 0) !== (rhsArgs[1]?.size ?? 0))
      return false;
    return true;
  }
}
