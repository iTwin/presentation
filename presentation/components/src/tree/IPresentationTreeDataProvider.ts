/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Hierarchies */

import { TreeDataProvider, TreeNodeItem } from "@bentley/ui-components/lib/tree/TreeDataProvider";
import IPresentationDataProvider from "../common/IPresentationDataProvider";
import { NodeKey, NodePathElement } from "@bentley/presentation-common";

/**
 * Presentation tree data provider.
 */
export default interface IPresentationTreeDataProvider extends TreeDataProvider, IPresentationDataProvider {
  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   */
  getNodeKey(node: TreeNodeItem): NodeKey;

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  getFilteredNodePaths(filter: string): Promise<ReadonlyArray<Readonly<NodePathElement>>>;
}
