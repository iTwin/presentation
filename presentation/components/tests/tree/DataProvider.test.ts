/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import "@bentley/presentation-frontend/tests/_helpers/MockFrontendEnvironment";
import { expect } from "chai";
import * as faker from "faker";
import * as moq from "@bentley/presentation-common/tests/_helpers/Mocks";
import { PromiseContainer } from "@bentley/presentation-common/tests/_helpers/Promises";
import { createRandomECInstanceNodeKey, createRandomECInstanceNode, createRandomNodePathElement } from "@bentley/presentation-common/tests/_helpers/random";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { Node, NodeKey } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import PresentationManager from "@bentley/presentation-frontend/lib/PresentationManager";
import { PageOptions, TreeNodeItem } from "@bentley/ui-components";
import PresentationTreeDataProvider from "../../lib/tree/DataProvider";
import { pageOptionsUiToPresentation } from "../../lib/tree/Utils";

describe("TreeDataProvider", () => {

    let rulesetId: string;
    let provider: PresentationTreeDataProvider;
    const presentationManagerMock = moq.Mock.ofType<PresentationManager>();
    const imodelMock = moq.Mock.ofType<IModelConnection>();
    before(() => {
        rulesetId = faker.random.word();
        Presentation.presentation = presentationManagerMock.object;
    });
    beforeEach(() => {
        presentationManagerMock.reset();
        provider = new PresentationTreeDataProvider(imodelMock.object, rulesetId);
    });

    const createTreeNodeItem = (key?: NodeKey, parentId?: string): TreeNodeItem => ({
        id: faker.random.uuid(),
        parentId,
        label: faker.random.word(),
        description: faker.random.words(),
        hasChildren: faker.random.boolean(),
        extendedData: { key: key || createRandomECInstanceNodeKey() },
    });

    describe("rulesetId", () => {

        it("returns rulesetId provider is initialized with", () => {
            expect(provider.rulesetId).to.eq(rulesetId);
        });

    });

    describe("connection", () => {

        it("returns connection provider is initialized with", () => {
            expect(provider.connection).to.eq(imodelMock.object);
        });

    });

    describe("getRootNodesCount", () => {

        it("returns presentation manager result", async () => {
            const result = faker.random.number();
            presentationManagerMock
                .setup((x) => x.getRootNodesCount({ imodel: imodelMock.object, rulesetId }))
                .returns(async () => result)
                .verifiable();
            const actualResult = await provider.getRootNodesCount();
            expect(actualResult).to.eq(result);
            presentationManagerMock.verifyAll();
        });

        it("memoizes result", async () => {
            const resultContainers = [new PromiseContainer<number>(), new PromiseContainer<number>()];
            presentationManagerMock
                .setup((x) => x.getRootNodesCount({ imodel: imodelMock.object, rulesetId }))
                .returns(() => resultContainers[0].promise);
            presentationManagerMock
                .setup((x) => x.getRootNodesCount({ imodel: imodelMock.object, rulesetId }))
                .returns(() => resultContainers[1].promise);
            const promises = [provider.getRootNodesCount(), provider.getRootNodesCount()];
            resultContainers.forEach((c: PromiseContainer<number>, index: number) => c.resolve(index));
            const results = await Promise.all(promises);
            expect(results[1]).to.eq(results[0]).to.eq(0);
            presentationManagerMock.verify((x) => x.getRootNodesCount({ imodel: imodelMock.object, rulesetId }), moq.Times.once());
        });

    });

    describe("getRootNodes", () => {

        it("returns presentation manager result", async () => {
            const pageOptions: PageOptions = { start: faker.random.number(), size: faker.random.number() };
            const result = [createRandomECInstanceNode(), createRandomECInstanceNode()];
            presentationManagerMock
                .setup((x) => x.getRootNodes({ imodel: imodelMock.object, rulesetId, paging: pageOptionsUiToPresentation(pageOptions) }))
                .returns(async () => result)
                .verifiable();
            const actualResult = await provider.getRootNodes(pageOptions);
            expect(actualResult).to.matchSnapshot();
            presentationManagerMock.verifyAll();
        });

        it("memoizes result", async () => {
            const resultContainers = [new PromiseContainer<Node[]>(), new PromiseContainer<Node[]>(), new PromiseContainer<Node[]>()];
            presentationManagerMock
                .setup((x) => x.getRootNodes({ imodel: imodelMock.object, rulesetId, paging: undefined }))
                .returns(() => resultContainers[0].promise)
                .verifiable(moq.Times.once());
            presentationManagerMock
                .setup((x) => x.getRootNodes({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 0 } }))
                .verifiable(moq.Times.never());
            presentationManagerMock
                .setup((x) => x.getRootNodes({ imodel: imodelMock.object, rulesetId, paging: { start: 1, size: 0 } }))
                .returns(() => resultContainers[1].promise)
                .verifiable(moq.Times.once());
            presentationManagerMock
                .setup((x) => x.getRootNodes({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 1 } }))
                .returns(() => resultContainers[2].promise)
                .verifiable(moq.Times.once());

            const promises = [
                provider.getRootNodes(undefined), provider.getRootNodes(undefined),
                provider.getRootNodes({ start: 0, size: 0 }), provider.getRootNodes({ start: 0, size: 0 }),
                provider.getRootNodes({ start: 1, size: 0 }), provider.getRootNodes({ start: 1, size: 0 }),
                provider.getRootNodes({ start: 0, size: 1 }), provider.getRootNodes({ start: 0, size: 1 }),
            ];
            resultContainers.forEach((c: PromiseContainer<Node[]>) => c.resolve([createRandomECInstanceNode()]));
            const results = await Promise.all(promises);

            expect(results[0]).to.eq(results[1], "results[0] should eq results[1]");
            expect(results[2])
                .to.eq(results[3], "results[2] should eq results[3]")
                .to.eq(results[0], "both results[2] and results[3] should eq results[0]");
            expect(results[4]).to.eq(results[5], "results[4] should eq results[5]");
            expect(results[6]).to.eq(results[7], "results[6] should eq results[7]");

            presentationManagerMock.verifyAll();
        });

    });

    describe("getChildNodesCount", () => {

        it("returns presentation manager result", async () => {
            const parentKey = createRandomECInstanceNodeKey();
            const parentNode = createTreeNodeItem(parentKey);
            const result = faker.random.number();
            presentationManagerMock
                .setup((x) => x.getChildrenCount({ imodel: imodelMock.object, rulesetId }, parentKey))
                .returns(async () => result)
                .verifiable();
            const actualResult = await provider.getChildNodesCount(parentNode);
            expect(actualResult).to.eq(result);
            presentationManagerMock.verifyAll();
        });

        it("memoizes result", async () => {
            const parentKeys = [createRandomECInstanceNodeKey(), createRandomECInstanceNodeKey()];
            const parentNodes = parentKeys.map((key) => createTreeNodeItem(key));
            const resultContainers = [new PromiseContainer<number>(), new PromiseContainer<number>()];

            presentationManagerMock
                .setup((x) => x.getChildrenCount({ imodel: imodelMock.object, rulesetId }, parentKeys[0]))
                .returns(() => resultContainers[0].promise)
                .verifiable(moq.Times.once());
            presentationManagerMock
                .setup((x) => x.getChildrenCount({ imodel: imodelMock.object, rulesetId }, parentKeys[1]))
                .returns(() => resultContainers[1].promise)
                .verifiable(moq.Times.once());

            const promises = [
                provider.getChildNodesCount(parentNodes[0]),
                provider.getChildNodesCount(parentNodes[1]),
                provider.getChildNodesCount(parentNodes[0]),
            ];
            resultContainers.forEach((c: PromiseContainer<number>, index: number) => c.resolve(index));
            const results = await Promise.all(promises);
            expect(results[0]).to.eq(results[2]).to.eq(0);
            expect(results[1]).to.eq(1);

            presentationManagerMock.verifyAll();
        });

    });

    describe("getChildNodes", () => {

        it("returns presentation manager result", async () => {
            const parentKey = createRandomECInstanceNodeKey();
            const parentNode = createTreeNodeItem(parentKey);
            const pageOptions: PageOptions = { start: faker.random.number(), size: faker.random.number() };
            presentationManagerMock
                .setup((x) => x.getChildren({ imodel: imodelMock.object, rulesetId, paging: pageOptionsUiToPresentation(pageOptions) }, parentKey))
                .returns(async () => [createRandomECInstanceNode(), createRandomECInstanceNode()])
                .verifiable();
            const actualResult = await provider.getChildNodes(parentNode, pageOptions);
            expect(actualResult).to.matchSnapshot();
            presentationManagerMock.verifyAll();
        });

        it("memoizes result", async () => {
            const parentKeys = [createRandomECInstanceNodeKey(), createRandomECInstanceNodeKey()];
            const parentNodes = parentKeys.map((key) => createTreeNodeItem(key));
            const resultContainers = [new PromiseContainer<Node[]>(), new PromiseContainer<Node[]>(), new PromiseContainer<Node[]>()];

            presentationManagerMock
                .setup((x) => x.getChildren({ imodel: imodelMock.object, rulesetId, paging: undefined }, parentKeys[0]))
                .returns(() => resultContainers[0].promise)
                .verifiable(moq.Times.once());
            presentationManagerMock
                .setup((x) => x.getChildren({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 0 } }, parentKeys[0]))
                .verifiable(moq.Times.never());
            presentationManagerMock
                .setup((x) => x.getChildren({ imodel: imodelMock.object, rulesetId, paging: { start: 1, size: 0 } }, parentKeys[0]))
                .returns(() => resultContainers[1].promise)
                .verifiable(moq.Times.once());
            presentationManagerMock
                .setup((x) => x.getChildren({ imodel: imodelMock.object, rulesetId, paging: undefined }, parentKeys[1]))
                .returns(() => resultContainers[2].promise)
                .verifiable(moq.Times.once());

            const promises = [
                provider.getChildNodes(parentNodes[0], undefined), provider.getChildNodes(parentNodes[0], undefined),
                provider.getChildNodes(parentNodes[0], { start: 0, size: 0 }), provider.getChildNodes(parentNodes[0], { start: 0, size: 0 }),
                provider.getChildNodes(parentNodes[0], { start: 1, size: 0 }), provider.getChildNodes(parentNodes[0], { start: 1, size: 0 }),
                provider.getChildNodes(parentNodes[1], undefined), provider.getChildNodes(parentNodes[1], undefined),
            ];
            resultContainers.forEach((c: PromiseContainer<Node[]>) => c.resolve([createRandomECInstanceNode()]));
            const results = await Promise.all(promises);

            expect(results[0]).to.eq(results[1], "results[0] should eq results[1]");
            expect(results[2])
                .to.eq(results[3], "results[2] should eq results[3]")
                .to.eq(results[0], "both results[2] and results[3] should eq results[0]");
            expect(results[4]).to.eq(results[5], "results[4] should eq results[5]");
            expect(results[6]).to.eq(results[7], "results[6] should eq results[7]");

            presentationManagerMock.verifyAll();
        });

    });

    describe("getFilteredNodes", () => {

        it("returns presentation manager result", async () => {
            const filter = faker.random.word();
            presentationManagerMock
                .setup((x) => x.getFilteredNodePaths({ imodel: imodelMock.object, rulesetId }, filter))
                .returns(async () => [createRandomNodePathElement(), createRandomNodePathElement()])
                .verifiable();
            const actualResult = await provider.getFilteredNodePaths(filter);
            expect(actualResult).to.matchSnapshot();
            presentationManagerMock.verifyAll();
        });

    });

});
