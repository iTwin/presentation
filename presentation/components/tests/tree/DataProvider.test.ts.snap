// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeDataProvider getChildNodes returns presentation manager result 1`] = `
Array [
  Object {
    "checkBoxState": 0,
    "description": "",
    "displayCheckBox": true,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "Specialist",
          "id": "0xdece000000df9b",
        },
        "pathFromRoot": Array [
          "1174cf69-1c86-4639-b762-a6c7177b413a",
          "a82a1d6b-7662-44db-84a0-2f30eb96766a",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": true,
    "id": "a82a1d6b-7662-44db-84a0-2f30eb96766a/1174cf69-1c86-4639-b762-a6c7177b413a",
    "isCheckBoxEnabled": false,
    "label": "orange",
    "labelBackColor": -1598776577,
    "labelBold": false,
    "labelForeColor": undefined,
    "labelItalic": false,
    "parentId": "5784aabc-4961-4ecf-a0d7-979d0b9c0ad3",
  },
  Object {
    "checkBoxState": 1,
    "description": "",
    "displayCheckBox": false,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "Manager",
          "id": "0x1196a00000172ed",
        },
        "pathFromRoot": Array [
          "5f4366dd-35f1-4bda-8dfc-430d1563a0ed",
          "148dd535-3dae-4e7e-b5a5-4a091ca80d9d",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": false,
    "id": "148dd535-3dae-4e7e-b5a5-4a091ca80d9d/5f4366dd-35f1-4bda-8dfc-430d1563a0ed",
    "isCheckBoxEnabled": false,
    "label": "Kroon",
    "labelBackColor": undefined,
    "labelBold": false,
    "labelForeColor": undefined,
    "labelItalic": false,
    "parentId": "5784aabc-4961-4ecf-a0d7-979d0b9c0ad3",
  },
]
`;

exports[`TreeDataProvider getFilteredNodes memoizes result 1`] = `
Array [
  Object {
    "children": Array [],
    "index": 793,
    "isMarked": true,
    "node": Object {
      "backColor": "rgb(
    198,
    240,
    197
  )",
      "description": undefined,
      "foreColor": undefined,
      "hasChildren": true,
      "isCheckboxEnabled": false,
      "isCheckboxVisible": false,
      "isChecked": true,
      "isEditable": false,
      "isExpanded": false,
      "isSelectable": false,
      "key": Object {
        "instanceKey": Object {
          "className": "Internal",
          "id": "0x2ab5000000e4a0",
        },
        "pathFromRoot": Array [
          "84b1c31e-f0c4-40ed-a875-21d56c69d45b",
          "013ca5dc-dac1-418d-a111-ebaaeec752ea",
        ],
        "type": "ECInstanceNode",
      },
      "label": "feed Czech Republic Tasty",
    },
  },
]
`;

exports[`TreeDataProvider getFilteredNodes memoizes result 2`] = `
Array [
  Object {
    "children": Array [],
    "index": 793,
    "isMarked": true,
    "node": Object {
      "backColor": "rgb(
    198,
    240,
    197
  )",
      "description": undefined,
      "foreColor": undefined,
      "hasChildren": true,
      "isCheckboxEnabled": false,
      "isCheckboxVisible": false,
      "isChecked": true,
      "isEditable": false,
      "isExpanded": false,
      "isSelectable": false,
      "key": Object {
        "instanceKey": Object {
          "className": "Internal",
          "id": "0x2ab5000000e4a0",
        },
        "pathFromRoot": Array [
          "84b1c31e-f0c4-40ed-a875-21d56c69d45b",
          "013ca5dc-dac1-418d-a111-ebaaeec752ea",
        ],
        "type": "ECInstanceNode",
      },
      "label": "feed Czech Republic Tasty",
    },
  },
]
`;

exports[`TreeDataProvider getFilteredNodes returns presentation manager result 1`] = `
Array [
  Object {
    "children": Array [],
    "index": 841,
    "isMarked": false,
    "node": Object {
      "backColor": "rgb(
    32,
    198,
    46
  )",
      "description": undefined,
      "foreColor": "#970E90",
      "hasChildren": false,
      "isCheckboxEnabled": false,
      "isCheckboxVisible": true,
      "isChecked": false,
      "isEditable": true,
      "isExpanded": false,
      "isSelectable": true,
      "key": Object {
        "instanceKey": Object {
          "className": "Taka",
          "id": "0xe0610000013f37",
        },
        "pathFromRoot": Array [
          "e4e25571-05b9-4034-8bb1-b7e5e34bee89",
          "3267e90a-8528-4b85-8ec0-2dd87fe4a098",
        ],
        "type": "ECInstanceNode",
      },
      "label": "Movies plum",
    },
  },
  Object {
    "children": Array [],
    "index": 992,
    "isMarked": true,
    "node": Object {
      "backColor": "rgb(
    238,
    33,
    44
  )",
      "description": "Et dolorem dolorum quasi quia hic sit libero deserunt non.",
      "foreColor": "#36C0D2",
      "hasChildren": true,
      "isCheckboxEnabled": true,
      "isCheckboxVisible": false,
      "isChecked": true,
      "isEditable": false,
      "isExpanded": false,
      "isSelectable": false,
      "key": Object {
        "instanceKey": Object {
          "className": "Extended",
          "id": "0x4081000000a62f",
        },
        "pathFromRoot": Array [
          "7c2dc083-a9d7-4f77-8e8f-8a394442664d",
          "12027c22-44b3-4f70-b5dc-2845797cdfdf",
        ],
        "type": "ECInstanceNode",
      },
      "label": "Mauritius Rupee lime North Carolina",
    },
  },
]
`;

exports[`TreeDataProvider getRootNodes returns presentation manager result 1`] = `
Array [
  Object {
    "checkBoxState": 1,
    "description": "",
    "displayCheckBox": true,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "invoice",
          "id": "0x7b87000000da88",
        },
        "pathFromRoot": Array [
          "024a20b7-2bb8-4494-a62e-693a6a8b8195",
          "c178f4ea-9c38-4231-81ba-f354563fe450",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": false,
    "id": "c178f4ea-9c38-4231-81ba-f354563fe450/024a20b7-2bb8-4494-a62e-693a6a8b8195",
    "isCheckBoxEnabled": false,
    "label": "payment Personal Loan Account",
    "labelBackColor": 258423039,
    "labelBold": false,
    "labelForeColor": undefined,
    "labelItalic": false,
    "parentId": undefined,
  },
  Object {
    "checkBoxState": 0,
    "description": "Sunt accusantium soluta nisi sunt ut.",
    "displayCheckBox": true,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "Frozen",
          "id": "0xb6170000007d71",
        },
        "pathFromRoot": Array [
          "ba59b9ca-c108-4e77-9c7a-76eb4fe14056",
          "4ac5919d-9414-4ede-8b7a-ce31c14ff29b",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": true,
    "id": "4ac5919d-9414-4ede-8b7a-ce31c14ff29b/ba59b9ca-c108-4e77-9c7a-76eb4fe14056",
    "isCheckBoxEnabled": true,
    "label": "Car invoice Incredible",
    "labelBackColor": undefined,
    "labelBold": false,
    "labelForeColor": 1309790719,
    "labelItalic": false,
    "parentId": undefined,
  },
]
`;
